core/src/main/java/org/apache/spark/io/ReadAheadInputStream.java

```diff
@@ -0,0 +1,292 @@
+    }
+  }
+
+  @Override
+  public synchronized int read() throws IOException {
+    int val = read(oneByte, 0, 1);
+    if (val == -1) {
+      return -1;
+    }
+    return oneByte[0] & 0xFF;
+  }
+
+  @Override
+  public synchronized int read(byte[] b, int offset, int len) throws IOException {
+    stateChangeLock.lock();
+    try {
+      len = readInternal(b, offset, len);
+    }
+    finally {
+      stateChangeLock.unlock();
+    }
+    return len;
+  }
+
+  /**
+   * Internal read function which should be called only from read() api. The assumption is that
+   * the stateChangeLock is already acquired in the caller before calling this function.
+   */
+  private int readInternal(byte[] b, int offset, int len) throws IOException {
+    assert (stateChangeLock.isLocked());
+    if (offset < 0 || len < 0 || offset + len < 0 || offset + len > b.length) {
+      throw new IndexOutOfBoundsException();
+    }
+    if (!activeBuffer.hasRemaining()) {
+      if (!isReadInProgress) {
+        // This condition will only be triggered for the first time read is called.
+        readAsync(activeBuffer);
+      }
+      waitForAsyncReadComplete();
+    }
+
+    if (isReadAborted) {
+      throw new IOException(readException);
+    }
+    if (isEndOfStream()) {
+      return -1;
+    }
+    len = Math.min(len, activeBuffer.remaining());
+    activeBuffer.get(b, offset, len);
+
+    if (activeBuffer.remaining() <= readAheadThresholdInBytes && !readAheadBuffer.hasRemaining()) {
+      readAsync(readAheadBuffer);
+    }
+    if (!activeBuffer.hasRemaining()) {
+      ByteBuffer temp = activeBuffer;
+      activeBuffer = readAheadBuffer;
+      readAheadBuffer = temp;
+    }
+    return len;
+  }
+
+  @Override
+  public synchronized int available() throws IOException {
+    stateChangeLock.lock();
+    // Make sure we have no integer overflow.
+    int val = (int) Math.min((long) Integer.MAX_VALUE,
+            (long) activeBuffer.remaining() + readAheadBuffer.remaining());
+    stateChangeLock.unlock();
+    return val;
+  }
+
+  @Override
+  public synchronized long skip(long n) throws IOException {
+    stateChangeLock.lock();
+    long skipped;
+    try {
+      skipped = skipInternal(n);
+    } finally {
+      stateChangeLock.unlock();
+    }
+    return skipped;
+  }
+
+  /**
+   * Internal skip function which should be called only from skip() api. The assumption is that
+   * the stateChangeLock is already acquired in the caller before calling this function.
+   */
+  private long skipInternal(long n) throws IOException {
+    assert (stateChangeLock.isLocked());
+    if (n <= 0L) {
+      return 0L;
+    }
+    if (isReadInProgress) {
+      waitForAsyncReadComplete();
+    }
+    if (available() >= n) {
+      // we can skip from the internal buffers
+      int toSkip = (int)n;
+      byte[] temp = new byte[toSkip];
+      while (toSkip > 0) {
+        int skippedBytes = read(temp, 0, toSkip);
+        toSkip -= skippedBytes;
+      }
```

Comment:
Do we need to actually `read` (which involves data copy) ? Simply moving limit's should suffice ?
Note: the suffix invariants in `readInternal` should be handled - namely scheduling async read and buffer flipping - which could be a moved into a method in itself.

Reply:
hmm actually that was making the code a little simpler at the expense of data copy. Changed it to avoid any data copy.